src/Kestrel.Core/Internal/Infrastructure/HttpCharacters.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
+{
+    internal static class HttpCharacters
+    {
+        private static readonly int _tableSize = 128;
+        private static readonly bool[] _alphaNumeric = InitializeAlphaNumeric();
+        private static readonly bool[] _authority = InitializeAuthority();
+        private static readonly bool[] _token = InitializeToken();
+        private static readonly bool[] _host = InitializeHost();
+        private static readonly bool[] _fieldValue = InitializeFieldValue();
+
+        internal static void Initialize()
+        {
+            // Access _alphaNumeric to initialize static fields
+            var initialize = _alphaNumeric;
```

Comment:
I've checked, and this class is now indeed marked beforefieldinit. I'm a little worried the JIT could optimize this way further delaying initialization. I think that's unlikely though. @benaadams do you know if that's possible?

Reply:
Could always pass the variable out and no-inline it so the Jit definitely can't optimise it away?
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
internal static bool[] Initialize()
{
    // Access _alphaNumeric to initialize static fields
    return _alphaNumeric;
}
```